src/AutoRest.CSharp/LowLevel/Output/OperationMethodChainBuilder.cs

```diff
@@ -0,0 +1,391 @@
+        {
+            var protocolMethodParameter = BuildParameter(inputParameter, frameworkParameterType ?? ChangeTypeForProtocolMethod(inputParameter.Type));
+
+            AddReference(name, inputParameter, protocolMethodParameter, SerializationBuilder.GetSerializationFormat(inputParameter.Type));
+            if (inputParameter.Kind != InputOperationParameterKind.Method)
+            {
+                return;
+            }
+
+            var convenienceMethodParameter = BuildParameter(inputParameter);
+            _orderedParameters.Add(new ParameterChain(convenienceMethodParameter, protocolMethodParameter, protocolMethodParameter));
+        }
+
+        private Parameter BuildParameter(in InputParameter operationParameter, CSharpType? typeOverride = null)
+        {
+            var type = typeOverride != null
+                ? typeOverride.WithNullable(operationParameter.Type.IsNullable)
+                : _typeFactory.CreateType(operationParameter.Type);
+
+            return Parameter.FromInputParameter(operationParameter, type, _typeFactory);
+        }
+
+        private void AddReference(string nameInRequest, InputParameter? operationParameter, Parameter parameter, SerializationFormat serializationFormat)
+        {
+            var reference = operationParameter != null ? CreateReference(operationParameter, parameter) : parameter;
+            _requestParts.Add(new RequestPartSource(nameInRequest, operationParameter, reference, serializationFormat));
+        }
+
+        private ReferenceOrConstant CreateReference(InputParameter? operationParameter, Parameter parameter)
+        {
+            if (operationParameter?.Kind == InputOperationParameterKind.Client)
+            {
+                var field = operationParameter.IsEndpoint ? _fields.EndpointField : _fields.GetFieldByParameter(parameter);
+                if (field == null)
+                {
+                    throw new InvalidOperationException($"Parameter {parameter.Name} should have matching field");
+                }
+
+                return new Reference(field.Name, field.Type);
+            }
+
+            if (operationParameter?.Kind == InputOperationParameterKind.Constant && parameter.DefaultValue != null)
+            {
+                return (ReferenceOrConstant)parameter.DefaultValue;
+            }
+
+            return parameter;
+        }
+
+        private CSharpType? ChangeTypeForProtocolMethod(InputType type) => type switch
+        {
+            InputEnumType enumType => _typeFactory.CreateType(enumType.EnumValueType).WithNullable(enumType.IsNullable),
+            InputModelType modelType => new CSharpType(typeof(object), modelType.IsNullable),
+            _ => null
+        };
+
+        private record ReturnTypeChain(CSharpType Convenience, CSharpType Protocol, CSharpType? ConvenienceResponseType);
+
+        private record ParameterChain(Parameter? Convenience, Parameter? Protocol, Parameter? CreateMessage);
```

Comment:
How about we call it `ParameterSet`? `ParameterSuit`?

Reply:
`ParameterGroup` will be misleading because we have grouped parameters in HLC. `ParameterSet` doesn't represent the semantics cause the idea that these parameters not just some random parameters brought together - they have a connection between them.